dbms/src/DataTypes/DataTypeDecimalBase.h

```diff
@@ -0,0 +1,225 @@
+///     S is maximum scale of operands. The allowed valuas are [0, precision]
+template <typename T>
+class DataTypeDecimalBase : public DataTypeWithSimpleSerialization
+{
+    static_assert(IsDecimalNumber<T>);
+
+public:
+    using FieldType = T;
+    using ColumnType = ColumnDecimal<T>;
+
+    static constexpr bool is_parametric = true;
+
+    static constexpr size_t maxPrecision() { return maxDecimalPrecision<T>(); }
+
+    DataTypeDecimalBase(UInt32 precision_, UInt32 scale_)
+    :   precision(precision_),
+        scale(scale_)
+    {
+        if (unlikely(precision < 1 || precision > maxPrecision()))
+            throw Exception("Precision " + std::to_string(precision) + " is out of bounds", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+        if (unlikely(scale < 0 || static_cast<UInt32>(scale) > maxPrecision()))
+            throw Exception("Scale " + std::to_string(scale) + " is out of bounds", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+    }
+
+    TypeIndex getTypeId() const override { return TypeId<T>::value; }
+
+    Field getDefault() const override;
+    MutableColumnPtr createColumn() const override;
+
+    bool isParametric() const override { return true; }
+    bool haveSubtypes() const override { return false; }
+    bool shouldAlignRightInPrettyFormats() const override { return true; }
+    bool textCanContainOnlyValidUTF8() const override { return true; }
+    bool isComparable() const override { return true; }
+    bool isValueRepresentedByNumber() const override { return true; }
+    bool isValueUnambiguouslyRepresentedInContiguousMemoryRegion() const override { return true; }
+    bool haveMaximumSizeOfValue() const override { return true; }
+    size_t getSizeOfValueInMemory() const override { return sizeof(T); }
+
+    bool isSummable() const override { return true; }
+    bool canBeUsedInBooleanContext() const override { return true; }
+    bool canBeInsideNullable() const override { return true; }
+
+    void serializeBinary(const Field & field, WriteBuffer & ostr) const override;
+    void serializeBinary(const IColumn & column, size_t row_num, WriteBuffer & ostr) const override;
+    void serializeBinaryBulk(const IColumn & column, WriteBuffer & ostr, size_t offset, size_t limit) const override;
+
+    void deserializeBinary(Field & field, ReadBuffer & istr) const override;
+    void deserializeBinary(IColumn & column, ReadBuffer & istr) const override;
+    void deserializeBinaryBulk(IColumn & column, ReadBuffer & istr, size_t limit, double avg_value_size_hint) const override;
+
+    /// Decimal specific
+
+    UInt32 getPrecision() const { return precision; }
+    UInt32 getScale() const { return scale; }
+    T getScaleMultiplier() const { return getScaleMultiplier(scale); }
+
+    T wholePart(T x) const
+    {
+        if (scale == 0)
+            return x;
+        return x / getScaleMultiplier();
+    }
+
+    T fractionalPart(T x) const
```

Comment:
Code duplication for most of functions in this file: here and in expracted DecimalFunctions.h

Reply:
fixed that, btw, there are only two functions that were "duplicating" code from `DataTypeDecimalBase.h`:
 * `decimalWholePart(x, scale);`
 * `decimalFractionalPart(x, scale);`